File: ValidAnagram.py
# Given two strings s and t, return true if t is an anagram of s, and false otherwise.

def isAnagram(s, t):
    if len(s) != len(t):
        return False

    ss = dict()
    st = dict()

    for i in s:
        if i not in ss:
            ss[i] = 1
        else:
            ss[i] += 1

    for i in t:
        if i not in st:
            st[i] = 1
        else:
            st[i] += 1

    return st == ss

#Slower Solution, a little more space efficient
def isAnagram2 (s, t):
    if len(s) != len(t):
        return False

    for val in s:
        if val in t:
            t = t.replace(val, "" ,1)
        else:
            return False
    return True

#Test cases
s = "anagram"
t = "nagaram"
print(isAnagram(s, t))
print(isAnagram2(s, t))

s = "rat"
t = "car"
print(isAnagram(s, t))
print(isAnagram2(s, t))

s = "aa"
t = "a"
print(isAnagram(s, t))
print(isAnagram2(s, t))

File: ValidAnagram.js
// Given two strings s and t, return true if t is an anagram of s, and false otherwise.
// An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

function isAnagram(s, t){
    sLetters = {};
    tLetters = {};

    if (new Set(s).size != new Set(t).size){
        return false; 
    }

    for (i = 0; i < s.length; i++){
        if (s[i] in sLetters){
            sLetters[s[i]] += 1;
        }else{
            sLetters[s[i]] = 1;
        }
    }

    for (i = 0; i < t.length; i++){
        if (t[i] in tLetters){
            tLetters[t[i]] += 1;
        }else{
            tLetters[t[i]] = 1;
        }
    }

    for ([key, val] of Object.entries(sLetters)){
        if (!(key in tLetters) || (tLetters[key] != val)){
            return false;
        }
    }
    return true;
}

// Test cases
s = "anagram"
t = "nagaram"
console.log(isAnagram(s, t))

s = "rat"
t = "car"
console.log(isAnagram(s, t))

s = "a"
t = "ab"
console.log(isAnagram(s, t))
